odoo/tools/populate.py

```diff
@@ -1,180 +1,403 @@
+                    table=SQL.identifier(model._table),
+                    src_columns=SQL(', ').join(SQL.identifier(field.name) for field in update_fields),
+                    dest_columns=SQL(', ').join(
+                        get_field_variation_char(field, postfix=SQL('CHR(%s)', separator_code))
+                        for field in update_fields))
+        model.env.cr.execute(query)
+    query = SQL("""
+        INSERT INTO %(table)s (%(dest_columns)s)
+        SELECT %(src_columns)s FROM %(table)s %(table_alias)s,
+        GENERATE_SERIES(1, %(factor)s) %(series_alias)s
+    """, table=SQL.identifier(model._table), factor=factors[model],
+         dest_columns=SQL(', ').join(dest_fields), src_columns=SQL(', ').join(src_fields),
+         table_alias=SQL.identifier(table_alias), series_alias=SQL.identifier(series_alias))
+    model.env.cr.execute(query)
+    # normally copying the 'id' will set the model entry in the populated dict,
+    # but for the case of a table with no 'id' (ex: Many2many), we add manually,
+    # by reading the key and having the defaultdict do the insertion, with a default value of 0
+    if populated[model]:
+        # in case we populated a model with an 'id', we update the sequence
+        update_sequence(model)
+
+
+def infer_many2many_model(env: Environment, field: Field) -> Model | object:
+    """
+    Returns the relation model used for the m2m:
+    - If it's a custom model, return the model
+    - If it's an implicite table generated by the ORM,
+      return a wrapped model that behaves like a fake duck-typed model for the population algorithm
+    """
+
+    class Many2oneFieldWrapper(Many2one):
+        def __init__(self, model, field_name, comodel_name):
+            super().__init__(comodel_name)
+            self._setup_attrs(model, field_name)  # setup most of the default attrs
+
+    class Many2manyModelWrapper:
+        def __init__(self, field):
+            self._name = field.relation  # a m2m doesn't have a _name, so we use the tablename
+            self._table = field.relation
+            self._inherits = {}
+            self.env = env
+            self._rec_name = None
+            self._rec_names_search = []
+            # if the field is inherited, the column attributes are defined on the base_field
+            column1 = field.column1 or field.base_field.column1
+            column2 = field.column2 or field.base_field.column2
+            # column1 refers to the model, while column2 refers to the comodel
+            self._fields = {
+                field.column1: Many2oneFieldWrapper(self, column1, field.model_name),
+                field.column2: Many2oneFieldWrapper(self, column2, field.comodel_name),
+            }
+
+        def __repr__(self):
+            return f"<Many2manyModelWrapper({self._name!r})>"
+
+        def __eq__(self, other):
+            return self._name == other._name
+
+        def __hash__(self):
+            return hash(self._name)
+
```

Comment:
You should move those classes outside the function.  In Python _class creation_ is quite slow, while _class instantiation_ is fast.

Reply:
I initially wanted to keep them scoped to the function, but it's true that each call would re-create the classes, I guess it's better to hoist them just above.